src/dotnet/commands/dotnet-build/IncrementalPreconditionManager.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.DotNet.Tools.Compiler;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Cli.Compiler.Common;
+using Microsoft.DotNet.ProjectModel.Utilities;
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    internal class IncrementalPreconditionManager
+    {
+        private readonly bool _printPreconditions;
+        private readonly bool _forceNonIncremental;
+        private readonly bool _skipDependencies;
+        private Dictionary<ProjectContextIdentity, IncrementalPreconditions> _preconditions;
+
+        public IncrementalPreconditionManager(bool printPreconditions, bool forceNonIncremental, bool skipDependencies)
+        {
+            _printPreconditions = printPreconditions;
+            _forceNonIncremental = forceNonIncremental;
+            _skipDependencies = skipDependencies;
+            _preconditions = new Dictionary<ProjectContextIdentity, IncrementalPreconditions>();
+        }
+
+        public static readonly string[] KnownCompilers = { "csc", "vbc", "fsc" };
+
+        public IncrementalPreconditions GetIncrementalPreconditions(ProjectGraphNode projectNode)
+        {
+            IncrementalPreconditions preconditions;
+            if (_preconditions.TryGetValue(projectNode.ProjectContext.Identity, out preconditions))
+            {
+                return preconditions;
+            }
+
+            preconditions = new IncrementalPreconditions(_printPreconditions);
+
+            if (_forceNonIncremental)
+            {
+                preconditions.AddForceUnsafePrecondition();
+            }
+
+            var projectsToCheck = GetProjectsToCheck(projectNode);
+
+            foreach (var project in projectsToCheck)
+            {
+                CollectScriptPreconditions(project, preconditions);
+                CollectCompilerNamePreconditions(project, preconditions);
+                CollectCheckPathProbingPreconditions(project, preconditions);
+            }
+            _preconditions[projectNode.ProjectContext.Identity] = preconditions;
+            return preconditions;
+        }
+
+        private List<ProjectContext> GetProjectsToCheck(ProjectGraphNode projectNode)
+        {
+            if (_skipDependencies)
+            {
+                return new List<ProjectContext>(1) { projectNode.ProjectContext };
+            }
+
+            // include initial root project
+            var contextsToCheck = new List<ProjectContext>(1 + projectNode.Dependencies.Count) { projectNode.ProjectContext };
+
+            // TODO: not traversing deeper than 1 level of dependencies
+            contextsToCheck.AddRange(projectNode.Dependencies.Select(n => n.ProjectContext));
```

Comment:
projectNode.Dependencies returns only the direct dependencies right? If yes, this should be transitive to avoid having mixed cones. Besides the mixed cones issues, PrintSummary in DotnetProjectBuilder won't show all the preconditions for all the nodes in the graph (it will only print dependencies for the root nodes and their direct children)